<b>Problem Statement 1: </b>Is gender independent of education level? A random sample of 395 people were surveyed and each person was asked to report the highest education level they
obtained. The data that resulted from the survey is summarized in the following table:
    
<table>
    <tr>
        <td></td>
        <td>High School</td>
        <td>Bachelors</td>
        <td>Masters</td>
        <td>Ph.d.</td>
        <td>Total</td>
    </tr>
    <tr>
        <td>Female</td>
        <td>60</td>
        <td>54</td>
        <td>46</td>
        <td>41</td>
        <td>201</td>
    </tr>
     <tr>
        <td>Male</td>
        <td>40</td>
        <td>44</td>
        <td>53</td>
        <td>57</td>
        <td>194</td>   
     </tr>
    <tr>
        <td>Total</td>
        <td>100</td>
        <td>98</td>
        <td>99</td>
        <td>98</td>
        <td>395</td>   
     </tr>
</table>

Question: Are gender and education level dependent at 5% level of significance? In other words, given the data collected above, is there a relationship between the gender of an individual and the level of education that they have obtained?


In [ ]:
# Solution



Problem assessment : It has been asked that , is there a relationship between is there a relationship between the gender of an individual and the level of education that they have obtained or that education level depends upon gender.        
In order to finding the relationship between gender and education level or whether both of two variable are dependent or independent , A Hypothesis testing can be performed on provided data using " Chi- Square test"


Definition : "Chi-Square test " or "Chi-Square Test of Independence" is a statistical test is used to determine whether there is a significant association between the two variables.The test is applied when you have two categorical variables from a single population. 

Formula : 

Chi-Square ( <b>Χ<sup> 2</sup></b> ) =<b> ∑(O − E )<sup> 2</sup> / E</b>

Where 
O = represents the observed frequency
E = is the expected frequency under the null hypothesis , and E can be evaluated by 

           E=( (row total) × (column total))/ (sample size)


In [ ]:
# State the Hypothesis

As per Chi-Square test of independece, the null hypothesis stats that there is no relationship exists on the categorical variables in the population; they are independent.

Hence for above mentioned problem :
    
(Null Hypothesis) Ho: The education level and gender independent to each other.

(Alternative Hypothesis)  Ha: The education level and gender dependent to each other.
    
    


In [1]:
# level of  significance (given in problem)
alpha= 0.05

In [2]:
# Degree of freedom : 

Note : In Chi-Square test , degree of freedom ( df) is , for a test of independence equals the product of the number of categories in each variable minus 1 .
       In above mentioned problem , there is 2 rows for Gender and 4 columns for eductaion level (except total colums).
       Hence degree of freedom (df) = (2-1)*(4-1) =3
    

In [3]:
df=3  #degree of freedom

In [4]:
sample_size =395 # sample size given in problem

In [5]:
# Let's gather the data 

list_Female=[60,54,46,41,201]  # temporary list for row of female
list_Male=[40,44,53,57,194]     # temporary list for row of male
list_Total=[100,98,99,98,395]    # temporary list for row of Total


comulative_list= list(zip(list_Female,list_Male,list_Total))  # zip function to map the idividual list at their index level
comulative_list

[(60, 40, 100), (54, 44, 98), (46, 53, 99), (41, 57, 98), (201, 194, 395)]

In [6]:
# transform the list into a pndas data farme 
import pandas as pd # pandas module import
df_table = pd.DataFrame(comulative_list).T  # " .T " to tranapose the data frame or convert columns into rows
df_table

0   1   2   3    4
0   60  54  46  41  201
1   40  44  53  57  194
2  100  98  99  98  395

In [7]:
# re-name dataframe  column and row-index

df_table.rename(columns={0:'High School',1:'Bachelors',2:'Masters',3:'Ph.d.',4:'Total'},index={0:'Female',1:'Male',2:'Total'} , inplace=True  )
df_table

High School  Bachelors  Masters  Ph.d.  Total
Female           60         54       46     41    201
Male             40         44       53     57    194
Total           100         98       99     98    395

In [8]:
# Chi- Square Test

# Formula :  Sum of (square of ( observation data- expected data) / ( expected data)) 


In [9]:
# calculate Observation data 

chi_observation= df_table.iloc[0:2,0:4]   # Slice the data frame to get data except "Total" from row and column, " iloc" to get data from position number
chi_observation  # Observed data

High School  Bachelors  Masters  Ph.d.
Female           60         54       46     41
Male             40         44       53     57

In [10]:
# calculate Expected data      i.e Sum of rows and sum of columns

df_table_row=df_table["Total"][0:3]  # Slice the data frame to get data "Total" from row 
df_table_col=df_table.loc["Total"][0:5]  # Slice the data frame to get data "Total" from column

print("Total's row data\n",df_table_row)
print("\nTotal's column data\n",df_table_col)

Total's row data
 Female    201
Male      194
Total     395
Name: Total, dtype: int64

Total's column data
 High School    100
Bachelors       98
Masters         99
Ph.d.           98
Total          395
Name: Total, dtype: int64


In [11]:
# expectd data :  which is equal >> Er,c = (nr * nc) / n

# where  , if A and B are two categorical variable  then , 
# Er,c is the expected frequency count for level r of Variable A and level c of Variable B,
# nr is the total number of sample observations at level r of Variable A, 
# nc is the total number of sample observations at level c of Variable B, 
# and n is the total sample size

import numpy as np # numpy for mathematical calculation

sample_size=395 # given in problem

chi_expected= (np.outer(df_table_row ,df_table_col))/sample_size  # np.outer () for cross multiplication of two same/differencet size of series/array
chi_expected

array([[ 50.88607595,  49.86835443,  50.37721519,  49.86835443,
        201.        ],
       [ 49.11392405,  48.13164557,  48.62278481,  48.13164557,
        194.        ],
       [100.        ,  98.        ,  99.        ,  98.        ,
        395.        ]])

In [12]:
# Convert expected data into data frame and assign column and row-index

chi_expected = pd.DataFrame(chi_expected) 

chi_expected.rename(columns={0:'High School',1:'Bachelors',2:'Masters',3:'Ph.d.',4:'Total'},index={0:'Female',1:'Male',2:'Total'}, inplace= True )

chi_expected  # Expected data

High School  Bachelors    Masters      Ph.d.  Total
Female    50.886076  49.868354  50.377215  49.868354  201.0
Male      49.113924  48.131646  48.622785  48.131646  194.0
Total    100.000000  98.000000  99.000000  98.000000  395.0

In [13]:

# Calculate chi- square stistics value

chi_squared_statistics=(((chi_observation-chi_expected)**2)/chi_expected).sum().sum() # Sum() has been applied twice once to get the column sums and a second time to add the column sums together, returning the sum of the entire 2D table
print(" Chi-Square Test Statstic value :", chi_squared_statistics )

 Chi-Square Test Statstic value : 8.006066246262538


In [14]:
# Calculate p-vale (probability) for Test statistics value

import scipy.stats as stats

df # df= degre of freedom , 3 for this problem, calcualtion has been done above.

p_value = 1 - stats.chi2.cdf(chi_squared_statistics,df)    #  stats.chi2.cdf is comulative density function which defines mass of probability of the function up to a given point; what percentage of the distribution lies on one side of this point. 
print("Chi-square P value")
print(p_value)

Chi-square P value
0.04588650089174717


In [16]:
#Calculate chi-square crtical value
alpha =0.05  # significance level , given in problem hence confidence level is 95 %


chi_squared_statistics_critical= stats.chi2.ppf(0.95,df)
print("Output:")
print(" Chi-Square Test Critical value :", chi_squared_statistics_critical )



Output:
 Chi-Square Test Critical value : 7.814727903251179


Result : 
The critical value of Chi-Sqaure(χ2) with 3 degree of freedom is 7.815. 

The test stattistics value of Chi-Sqaure(χ2) with 3 degree of freedom is 8.006. 

Since 8.006 > 7.815, therefore we reject the null hypothesis (Ho) and conclude that the education level depends on gender at a 5% level of significance.

-----------------------------------------------------------------------------------------------------------------------

<b>Problem Statement 2:</b> Using the following data, perform a oneway analysis of variance using α=.05.
Write up the results in APA format.

[Group1: 51, 45, 33, 45, 67]

[Group2: 23, 43, 23, 43, 45]

[Group3: 56, 76, 74, 87, 56]

In [17]:
# Solution

Analysis of variance (ANOVA) is a statistical technique that is used to check if the means of two or more groups are significantly different from each other. ANOVA checks the impact of one or more factors by comparing the means of different samples.



One way ANOVA : A one way ANOVA is used to compare two means from two independent (unrelated) groups using the F-distribution. The null hypothesis for the test is that the two means are equal. Therefore, a significant result means that the two means are unequal.

The whole idea behind the analysis of variance is to compare the ratio of between group variance to within group variance. If the variance caused by the interaction between the samples is much larger when compared to the variance that appears within each group, then it is because the means aren't the same.

Limitations of one-way ANOVA
A one-way ANOVA tells us that at least two groups are different from each other. But it won’t tell us which groups are different. If  test returns a significant f-statistic, we may need to run a post-hoc test to tell us exactly which groups have a difference in means. Below I have mentioned the steps to perform one-way ANOVA in Excel along with a post-hoc test.





Stages of calculating the One way Anova : (ref link : https://people.richland.edu/james/lecture/m170/ch13-1wy.html)

1. Hypothesis
2. Grand Mean Calculation
3. Between Goup Variation
4. Within Group Variation 
5. Total variation calculation
6. F test statistic
7. Summary Table


In [18]:
# Stage -1 : State Hypothesis

One way ANOVA , also has two type of Hypothesis , Null Hypothesis (H0) and Alternative Hypothesis (H1).
The Null hypothesis in ANOVA is valid when all the sample means are equal, or they don’t have any significant difference. Thus, they can be considered as a part of a larger set of the population. On the other hand, the alternate hypothesis is valid when at least one of the sample means is different from the rest of the sample means.


Thus  

H0 = <b> μ1 = μ2 = μ3 .... = ul </b>   Null hypothesis

H1 = <b>μl != μm</b>                   Alternative Hypothesis 

where μl , μm   belong to any two sample means out of all the samples considered for the test. 

In [19]:
# Level of significance

alpha =0.05

In [22]:
# Stage-2 : Grand Mean Calculation 

In [31]:
#Let's gather the data

group_1 = [51, 45, 33, 45, 67]   # temp list
group_2 = [23, 43, 23, 43, 45]   # temp list
group_3 = [56, 76, 74, 87, 56]   # temp list

print("Group_1 ",group_1)
print("Group_2 ",group_2)
print("Group_3 ",group_3)

Group_1  [51, 45, 33, 45, 67]
Group_2  [23, 43, 23, 43, 45]
Group_3  [56, 76, 74, 87, 56]


In [32]:
# Calulate the mean of mentioned data 

import statistics  # Statistics module for statistical calculation

mean_group_1 = statistics.mean(group_1)
mean_group_2 = statistics.mean(group_2)
mean_group_3 = statistics.mean(group_3)
print("Mean of group 1:",mean_group_1)
print("Mean of group 2:",mean_group_2)
print("Mean of group 3:",mean_group_3)

Mean of group 1: 48.2
Mean of group 2: 35.4
Mean of group 3: 69.8


In [33]:
# Calculate the Grand mean : 

grand_mean = statistics.mean([mean_group_1,mean_group_2,mean_group_3])
print("Grand Mean :" ,round(grand_mean,2))

Grand Mean : 51.13


In [30]:
# Stage -3 Between-group variability or deviation between group

<b>Between-group variability:</b> It refers to variations between the distributions of individual groups (or levels) as the values within each group are different. Each sample is looked at and the difference between its mean and grand mean is calculated to calculate the variability.

In [34]:
# devtion between Groups  = mean of groups - grand mean

dev_between_group_1= mean_group_1 - grand_mean
dev_between_group_2= mean_group_2 - grand_mean
dev_between_group_3= mean_group_3 - grand_mean
print("deviation between group 1 from grand mean: ",dev_between_group_1)
print("deviation between group 2 from grand mean: ",dev_between_group_2)
print("deviation between group 3 from grand mean: ",dev_between_group_3)


deviation between group 1 from grand mean:  -2.93333333333333
deviation between group 2 from grand mean:  -15.733333333333334
deviation between group 3 from grand mean:  18.666666666666664


In [35]:
# square of devation between groups

# gettting square of values for deviation between groups 
sq_dev_between_group= [x **2  for  x in [dev_between_group_1,dev_between_group_2,dev_between_group_3]] # list comprehension
sq_dev_between_group

[8.604444444444425, 247.53777777777782, 348.44444444444434]

In [36]:
sum_of_sq_dev_between_group= sum(sq_dev_between_group) # " Sum method for add-on numerical list elements"
print("Sum of Squared vales of devaition bvetween groups", sum_of_sq_dev_between_group)

Sum of Squared vales of devaition bvetween groups 604.5866666666666


In [38]:
# Calculate the between Group variation

# Note:  There are 3 groups and each group has 5 elements Thus
each_group_sample_size=5
no_of_groups=3
#  degree of freedom = Total No. of groups -1
degree_of_freedom_between_group= no_of_groups-1

between_group_variation= sum_of_sq_dev_between_group / degree_of_freedom_between_group 
print("Variation between groups",between_group_variation)

Variation between groups 302.2933333333333


In [39]:

# Mean sqaure betwen groups has been takes as product of variation between gropus and sample size of each group
mean_Square_Between_groups =between_group_variation *each_group_sample_size 
print("mean square between groups",mean_Square_Between_groups)

mean square between groups 1511.4666666666665


In [40]:
# Sum of sqaure betwen groups has been takes as product of Mean sqaure betwen groups and The degree of freedom between groups

sum_of_Squares_Between_groups = mean_Square_Between_groups* degree_of_freedom_between_group
print("sum-of-squares for between-group variability ",sum_of_Squares_Between_groups)

sum-of-squares for between-group variability  3022.933333333333


In [ ]:
# Stage -4 : Within-group variation or Deveiation betwee group elements

<b>Within-group variation</b>: It refers to variations caused by differences within individual groups (or levels) as not all the values within each group are the same. Each sample is looked at on its own and variability between the individual points in the sample is calculated. In other words, no interactions between samples are considered.

Within-group variability by looking at how much each value in each sample differs from its respective sample mean.


In [41]:
# Calculate Within-group variation

# stpe-1 : deviation wihin group: calculation done as difference between  mean of individual group and each group elements

dev_within_group_1 = [round((mean_group_1-x),3) for x in group_1]
dev_within_group_2 = [round((mean_group_2-x),3) for x in group_2]
dev_within_group_3 = [round((mean_group_3-x),3) for x in group_3]


print(" Deviation with-in group 1:",dev_within_group_1)
print(" Deviation with-in group 2:",dev_within_group_2)
print(" Deviation with-in group 2:",dev_within_group_3)


 Deviation with-in group 1: [-2.8, 3.2, 15.2, 3.2, -18.8]
 Deviation with-in group 2: [12.4, -7.6, 12.4, -7.6, -9.6]
 Deviation with-in group 2: [13.8, -6.2, -4.2, -17.2, 13.8]


In [44]:
# step -2  square deviation with group

sq_dev_within_group_1 = [round(x**2,3) for x in dev_within_group_1]
sq_dev_within_group_2 = [round(x**2,3) for x in dev_within_group_2]
sq_dev_within_group_3 = [round(x**2,3) for x in dev_within_group_3]

print(" Square of Deviation with-in group 1:",sq_dev_within_group_1)
print(" Square of Deviation with-in group 2:",sq_dev_within_group_2)
print(" Square of Deviation with-in group 3:",sq_dev_within_group_3)

 Square of Deviation with-in group 1: [7.84, 10.24, 231.04, 10.24, 353.44]
 Square of Deviation with-in group 2: [153.76, 57.76, 153.76, 57.76, 92.16]
 Square of Deviation with-in group 3: [190.44, 38.44, 17.64, 295.84, 190.44]


In [46]:
# step 3 sum of square deviation with in group

sum_of_sq_dev_within_group_1 =round(sum(sq_dev_within_group_1),2)
sum_of_sq_dev_within_group_2 =round(sum(sq_dev_within_group_2),2)
sum_of_sq_dev_within_group_3 =round(sum(sq_dev_within_group_3),2)


print(" sum of  Squared Deviation  with-in group 1:",sum_of_sq_dev_within_group_1)
print(" sum of  Squared Deviation  with-in group 2:",sum_of_sq_dev_within_group_2)
print(" sum of  Squared Deviation  with-in group 3:",sum_of_sq_dev_within_group_3)



 sum of  Squared Deviation  with-in group 1: 612.8
 sum of  Squared Deviation  with-in group 2: 515.2
 sum of  Squared Deviation  with-in group 3: 732.8


In [47]:
# each group has 5 elements hence , length of gropu is 5 and degree of freedom is :

No_of_elements_in_each_group=5
degree_of_freedom_within_indvidual_grp =No_of_elements_in_each_group-1


within_group_variation_grp1= sum_of_sq_dev_within_group_1/degree_of_freedom_within_indvidual_grp
within_group_variation_grp2= sum_of_sq_dev_within_group_2/degree_of_freedom_within_indvidual_grp
within_group_variation_grp3= sum_of_sq_dev_within_group_3/degree_of_freedom_within_indvidual_grp


print(" Variation with in group 1:",within_group_variation_grp1)
print(" Variation with in group 2:",within_group_variation_grp2)
print(" Variation with in group 3:",within_group_variation_grp3)

 Variation with in group 1: 153.2
 Variation with in group 2: 128.8
 Variation with in group 3: 183.2


In [48]:

# Mean sqaure with-in group has been calculated as mean of all with-in group variations

mean_Square_within_groups = (within_group_variation_grp1+within_group_variation_grp2+within_group_variation_grp3)/ 3 # divide by 3 has been done there are total 3 with-in group variation
print("Mean square with in group ",mean_Square_within_groups)

Mean square with in group  155.06666666666666


In [50]:
no_of_samples =3 # as 3 groups mentioned
sum_of_length_of_grp = 5+5+5 # as each gropu has 5 elements

# degree of freedom has been calculated as difference between total length of all groups/samples and no.of gropus /samples has been taken inot consideration
degree_of_freedom_within_grp =sum_of_length_of_grp - no_of_samples

# Sum of sqaures for with-in group has been calculated as product of Mean square with in group and degree of freedom with in group
sum_of_Squares_within_groups = mean_Square_within_groups* (degree_of_freedom_within_grp)
print("sum of squares for within-group variability:",sum_of_Squares_within_groups)

sum of squares for within-group variability: 1860.8


In [54]:
# Stage -5 Total variation : in Anova Test is , Sum of squares of with in group variability and between group variability

In [55]:
# Total Variation

total_variation=  sum_of_Squares_Between_groups+sum_of_Squares_within_groups 
print("Total variation is ",total_variation)

Total variation is  4883.733333333333


In [56]:
# Stage 6: F- test statistics : F = Between group variability / Within group variability

In [84]:
# F-Statistics 

F_test_Statistics=  mean_Square_Between_groups/mean_Square_within_groups
print("F-test statitics is", F_test_Statistics)

F-test statitics is 9.747205503009457


In [89]:
# p- value Calculation 
import scipy.stats

p_value = scipy.stats.f.sf(F_test_Statistics, 2, 12) #" scipy.stats.f.sf() for p-value calcuateion based upon f-test statistics"
print("p-value is:",p_value)

p-value is: 0.003059754143443061


In [91]:
# F-Critical 

print("Degree of freedom between in group:",degree_of_freedom_between_group)

print("Degree of freedom with in group:",degree_of_freedom_within_grp)


Degree of freedom between in group: 2
Degree of freedom with in group: 12


In [92]:
import scipy.stats # Scipy.stats module for statistical calculation
alpha = 0.05  

F_critical = scipy.stats.f.ppf(1-alpha, 2 ,12)

print("F-critical is :",F_critical)

F-critical is : 3.8852938346523933


<b>ANOVA Result</b> : Thus, F_test_Statistics > F_critical or 9.747205 > 3.88529, hence we reject null hypothesis that mean of samples are equal. Another way of examination is p-value for F-Stattistics is 0.003 which is less than alpha(level of significance=0.05), thus reject the null hypothesis

Effect size, in ANOVA, is a value which results how much  independent variable (like: sum of sqauared between group variation) has affected the dependent variable (Total veriation between group) in an experimental study

In [93]:
# Effect size

η2=sum_of_Squares_Between_groups/total_variation

print("Measure of effect size(η2):",η2)


Measure of effect size(η2): 0.61898001528885


<b>APA writeup </b>

F(2, 12)=9.75, p(0.003) <0.05, η2=0.62

<b>ANOVA Table</b>

<table>


<tr>
    <td><b>Source</b></td>
    <td><b>SS(Sum of Suqared)</b></td>
    <td><b>df(degree of freedom)</b></td>
    <td><b>MS(Mean Squared)</b></td>
    <td><b>F-Statistic</b></td>
    <td><b>p-value</b></td>
    <td><b>F-critical</b></td>
    <td><b>α-alpha level</b></td>
</tr>
<tr>    
    <td>Group( between group variation)</td>
    <td>3022.9</td>
    <td>2</td>
    <td>1511.45</td>
    <td>9.75</td>
    <td>0.003</td>
    <td>3.885</td>
    <td>0.05</td>
</tr>
 <tr>
    <td>Error( with-in group variation)</td>
    <td>1860.8</td>
    <td>12</td>
    <td>155.07</td>
 </tr>
 <tr>
    <td>Total</td>
    <td>4883.7</td>
 </tr>
</table>

-----------------------------------------------------------------------------------------------------------------------------

<b> Problem Statement 3: </b>
Calculate F Test for given 10, 20, 30, 40, 50 and 5,10,15, 20, 25.

For 10, 20, 30, 40, 50:

In [94]:
# Solution 

Definition : F-Test is a test statistic which has an F-distribution under the null hypothesis is called an F test. It is used to compare statistical models as per the data set provided or available.
F Test is generally defined as ratio of the variances of the given two set of values.

Formula : F-test 

F Value = <b>variance1 / variance2 = σ1<sup> 2</sup> /σ2<sup> 2</sup></b>


where :
σ<sup>2</sup>  = variance = square of standard devaition of values


In [95]:
# Let's gather the data 
list_1= [10, 20, 30, 40, 50]  # temp list
list_2= [5,10,15, 20, 25] # temp list


In [96]:
# convert lists into array using numpy module

import numpy as np

array_1=np.array(list_1)
array_2=np.array(list_2)
print("Input data")
array_1,array_2

Input data


(array([10, 20, 30, 40, 50]), array([ 5, 10, 15, 20, 25]))

In [98]:
# immport staticsc module for variance analysis

import statistics  

variance_1=statistics.variance(array_1)
variance_2=statistics.variance(array_2)
print("Variance for [10, 20, 30, 40, 50] is : ",variance_1)
print("Variance for [ 5, 10, 15, 20, 25] is:  ",variance_2)


Variance for [10, 20, 30, 40, 50] is :  250
Variance for [ 5, 10, 15, 20, 25] is:   62


In [99]:
# F-test calculation
F_Test= variance_1/variance_2
print("Output")
print("F-Test value is :",round(F_Test))  # round function to get the values till the decimal point

Output
F-Test value is : 4.0


In [100]:
# F-test Calculate with it's p-value using  scipy.stats.f_oneway

import scipy.stats
F_Test_oneway , p_value = scipy.stats.f_oneway(array_1,array_2)
print("F-Test value is :",round(F_Test_oneway))
print("p-value from the F-distribution :",round(p_value,3))

F-Test value is : 4.0
p-value from the F-distribution : 0.094
